File: easy/array/isomorphic_strings/isomorphic_strings.go
package main

import "fmt"

func isIsomorphic(s string, t string) bool {
	sTot, tTos := make(map[rune]rune, len(s)), make(map[rune]rune, len(s))

	for i, sChar := range s {
		tCharFromMap, ok1 := sTot[sChar]
		sCharFromMap, ok2 := tTos[rune(t[i])]
		if !ok1 && !ok2 {
			sTot[sChar] = rune(t[i])
			tTos[rune(t[i])] = sChar
		} else if !ok1 || !ok2 {
			return false
		} else if tCharFromMap != rune(t[i]) || sCharFromMap != sChar {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println("Solution 1 : ", isIsomorphic("egg", "add"))
	fmt.Println("Solution 2 : ", isIsomorphic("foo", "bar"))
	fmt.Println("Solution 3 : ", isIsomorphic("paper", "title"))
}

File: easy/array/isomorphic_strings/isomorphic_strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        s_to_t, t_to_s = {}, {}
        for i, s_char in enumerate(s):
            if not s_char in s_to_t and not t[i] in t_to_s:
                s_to_t[s_char] = t[i]
                t_to_s[t[i]] = s_char
            elif not s_char in s_to_t or not t[i] in t_to_s:
                return False
            elif t_to_s[t[i]] != s_char or s_to_t[s_char] != t[i]:
                return False

        return True


s = Solution()
print("Solution 1 : ", s.isIsomorphic("egg", "add"))
print("Solution 2 : ", s.isIsomorphic("foo", "bar"))
print("Solution 3 : ", s.isIsomorphic("paper", "title"))
